### PR TITLE
#519 Add setValue methods to Value types

### PR DESCRIPTION
**Pull request by MikhailErofeev**
add ResettableValue interface to base values and refactor setValue() in StringValue

### CODE REVIEW

**Comment by MikhailErofeev** [comment]
oops, Travis gone red at 'stratosphere-tests', with strange error. 

"eu.stratosphere.client.program.ProgramInvocationException: The program execution failed: java.lang.NoSuchMethodError: eu.stratosphere.types.IntValue.setValue(I)V
    at eu.stratosphere.examples.scala.datamining.KMeansForTest$UDTSerializerImpl$2.serialize(KMeansForTest.scala:79)
    at eu.stratosphere.examples.scala.datamining.KMeansForTest$UDTSerializerImpl$2.serialize(KMeansForTest.scala)
    at eu.stratosphere.examples.scala.datamining.KMeansForTest$$anon$8.readRecord(KMeansForTest.scala:79)
    at eu.stratosphere.examples.scala.datamining.KMeansForTest$$anon$8.readRecord(KMeansForTest.scala:79)
    at eu.stratosphere.api.common.io.DelimitedInputFormat.nextRecord(DelimitedInputFormat.java:449)
    at eu.stratosphere.pact.runtime.task.DataSourceTask.invoke(DataSourceTask.java:151)
    at eu.stratosphere.nephele.execution.RuntimeEnvironment.run(RuntimeEnvironment.java:345)
    at java.lang.Thread.run(Thread.java:744)"

why KMeansForTest commented? what is the UDTSerializerImpl? where are in code invocation of  setValue() ?

**Comment by fhueske** [comment]
Thanks a lot for your pull request and welcome!

The Scala API automatically generates data structures (List, Records, etc.) based on the basic data types (IntValue, StringValue, ...). The UDTSerializerImpl is auto-generated code to read and write one of these data structures if I am not mistaken.

I think we should keep the setValue methods for the primitive types (`setValue(int i)`, `setValue(double d)`, ...) as well for performance reasons. Otherwise, one would need to create (or let autoboxing create) an object just to update a value. Since these methods tend to be called very often, this could increase the pressure on garbage collection which we try to avoid as much as possible. Adding the methods for the basic types should fix the issue with Scala data type deserializer.

Also, the issue #519 did mean to add methods like `setValue(IntValue i)` and `setValue(DoubleVal)` such that the value of an IntValue can be set using another IntValue.

Best, Fabian

**Comment by StephanEwen** [comment]
The error comes from a special test whose classes are in a jar file under
/src/test/resources. We have them not in the normal compile/test scope to
verify the behavior of classloading with user-defined classes; that is why
the test is commented. If you change fundamental signatures, you need to
rebuild that test.

**Comment by MikhailErofeev** [comment]
oops, i misunderstand issue, and implement it to java primitives wrappers instead of Value. fix it

**Comment by fhueske** [comment]
OK, thanks a lot! :-)

File: stratosphere-core/src/main/java/eu/stratosphere/types/StringValue.java [hunk 1]
```diff
@@ -137,17 +138,7 @@ public String getValue() {
 	 * @param value The new string value.
 	 */
 	public void setValue(CharSequence value) {
-		if (value == null)
```
Comment (StephanEwen) [inline stratosphere-core/src/main/java/eu/stratosphere/types/StringValue.java:140]: I would suggest to keep this check, it gives a more specific error message.

**Comment by StephanEwen** [comment]
Looks good. One question though: You removed a few checks for null. The old checks gave a specific error message. When now someone submits a null value, the `value.value` statements will all throw an unspecific null pointer exception.

**Comment by MikhailErofeev** [comment]
ok, I replace this with apache Validate, forgot about old checking, then thought that exception will occurs few lines below and remove checking. Should i add validators in other setters?

**Comment by StephanEwen** [comment]
I think it looks good. I will merge it.

**Comment by StephanEwen** [comment]
Merged in 2006980be27debb604b40dc8f5c0ea229a4ae0ac

Thanks!

**Pull request closed without merging**